internal/server/openapi.go

```diff
@@ -34,33 +34,37 @@ var (
 	}
 )
 
-// register an API endpoint that has both a request and response value.
-func register[Req, Res any](a *API, method, path string, handler ReqResHandlerFunc[Req, Res]) {
-	funcName := getFuncName(handler)
-
-	//nolint:gocritic
-	rqt := reflect.TypeOf(*new(Req))
-	//nolint:gocritic
-	rst := reflect.TypeOf(*new(Res))
-
-	a.register(method, path, funcName, rqt, rst)
+type routeDefinition struct {
+	Method   string
+	Path     string
+	Name     string
+	Request  reflect.Type
+	Response reflect.Type
 }
 
-// registerDelete registers an API endpoint that has no response value, which is
-// currently only endpoints that use the DELETE method.
-func registerDelete[Req any](a *API, method, path string, handler ReqHandlerFunc[Req]) {
-	funcName := getFuncName(handler)
+func newDefinition(method string, path string, handler interface{}) routeDefinition {
+	fn := reflect.TypeOf(handler)
+	if fn.Kind() != reflect.Func {
+		panic("API routes must be functions")
+	}
 
-	//nolint:gocritic
-	rqt := reflect.TypeOf(*new(Req))
-	rst := reflect.TypeOf(nil)
+	resp := reflect.TypeOf(nil)
+	if fn.NumOut() == 2 { // the last out is always type error
+		resp = fn.Out(0)
+	}
 
-	a.register(method, path, funcName, rqt, rst)
+	return routeDefinition{
+		Method:   method,
+		Path:     path,
+		Name:     getFuncName(handler),
+		Request:  fn.In(1),
```

Comment:
what is `1` here?

Reply:
I should comment this. `1` is the second argument to the request function, the first argument is always `gin.Context`